pkg/kube/log_reader.go

```diff
@@ -0,0 +1,41 @@
+package kube
+
+import (
+	"io"
+
+	"k8s.io/client-go/rest"
+)
+
+var _ io.ReadCloser = (*logReader)(nil)
+
+func newLogReader(rw rest.ResponseWrapper) *logReader {
+	return &logReader{rw: rw}
+}
+
+// logReader defers a call to ResponseWrapper.Stream() until Read is called.
+// This is to help handle kubernetes behavior where calls to
+// Pod.GetLogs(...).Stream() will hang until at least one byte is returned.
+// kubectl logs handles this in
+// https://github.com/kubernetes/kubernetes/pull/67573/files#diff-12d472fe036bbe778e84dffc71564eb1R355
+type logReader struct {
+	rw  rest.ResponseWrapper
+	err error
+	rc  io.ReadCloser
+}
+
+func (lr *logReader) Read(p []byte) (n int, err error) {
+	if lr.rc == nil {
+		lr.rc, lr.err = lr.rw.Stream()
```

Comment:
A potential downside here is that the connection is not pre-established. But that may be OK for the use case.

Reply:
Also, this has an "interesting" behavior.
If the first read fails, I may be able to call `Read()` again, and have it succeed later, which is surprising.
An error from `rc.Read()` is not cached, but I'm guessing the expectation is that if `rc.Read()` fails once, then it will fail from then on.

Reply:
I agree it would be less surprising to return an error if it was cached. I'll change the check to:
```
if lr.rc == nil || lr.err != nil {
```